src/services/codefixes/inferFromUsage.ts

```diff
@@ -201,15 +215,70 @@ namespace ts.codefix {
     }
 
     function annotateJSDocParameters(changes: textChanges.ChangeTracker, sourceFile: SourceFile, parameterInferences: ParameterInference[], program: Program, host: LanguageServiceHost): void {
-        const result = mapDefined(parameterInferences, inference => {
+        const signature = parameterInferences.length && parameterInferences[0].declaration.parent;
+        if (!signature) {
+            return;
+        }
+        const paramTags = createUnderscoreEscapedMap<JSDocParameterTag>();
+        forEach(parameterInferences, inference => {
             const param = inference.declaration;
-            const typeNode = inference.type && getTypeNodeIfAccessible(inference.type, param, program, host);
-            return typeNode && !param.initializer && !getJSDocType(param) ? { ...inference, typeNode } : undefined;
+            // only infer parameters that have (1) no type and (2) an accessible inferred type
+            if (!param.initializer && !getJSDocType(param) && isIdentifier(param.name) && inference.type) {
+                const typeNode = getTypeNodeIfAccessible(inference.type, param, program, host);
+                if (typeNode) {
+                    paramTags.set(param.name.escapedText, createJSDocParamTag(typeNode, param.name, !!inference.isOptional, ""));
+                }
+            }
         });
-        changes.tryInsertJSDocParameters(sourceFile, result);
+        insertMergedJSDoc(
+            paramTags, changes, sourceFile, signature,
+            t => isJSDocParameterTag(t) && isIdentifier(t.name) && t.name.escapedText,
+            (synthetic, existing) => {
+                synthetic.comment = existing.comment;
+                synthetic.isBracketed = existing.isBracketed;
+            });
+    }
+
+    /**
+     * @param getKey When getKey returns a string, it is also asserting that `t`, the existing tag whose key is returned,
+     * is of type T, where T is the type of the new, synthetic tags.
+     */
+    function insertMergedJSDoc<T extends JSDocTag>(
+        syntheticTags: UnderscoreEscapedMap<T>,
+        changes: textChanges.ChangeTracker,
+        sourceFile: SourceFile,
+        parent: HasJSDoc,
+        getKey: (tag: JSDocTag) => __String | false,
+        merge: (synthetic: T, existing: T) => void) {
+        const comments = [];
+        const tags: JSDocTag[] = [];
+        if (hasJSDocNodes(parent)) {
+            for (const jsdoc of parent.jsDoc!) {
+                if (jsdoc.comment) {
+                    comments.push(jsdoc.comment);
+                }
+                if (jsdoc.tags) {
+                    for (const existing of jsdoc.tags) {
+                        const key = getKey(existing);
+                        const synthetic = key && syntheticTags.get(key);
+                        if (synthetic) {
+                            merge(synthetic, existing as T);

```

Comment:
This might be simpler as `tags.push(merge(synthetic, existing as T));` where we create a new synthetic node instead of mutating an existing one -- I believe there were plans to make nodes immutable in the future.

Reply:
Also might want to comment the reason we push to `tags` here instead of doing it at the bottom -- looks like we're trying to preserve the original order of tags as much as possible.